### PR TITLE
IGNITE-17328 Introduce implicit transactions from SQL

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java [hunk 1]
```diff
@@ -173,16 +188,79 @@ public void ddl() throws ExecutionException, InterruptedException {
     public void dml() throws ExecutionException, InterruptedException {
         sql("CREATE TABLE TEST(ID INT PRIMARY KEY, VAL0 INT)");
 
-        IgniteSql sql = CLUSTER_NODES.get(0).sql();
+        IgniteSql sql = igniteSql();
         Session ses = sql.createSession();
 
+        SqlQueryProcessor qryProc = (SqlQueryProcessor) IgniteTestUtils.getFieldValue(sql, IgniteSqlImpl.class, "qryProc");
```
Comment (ygerzhedovich) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java:194]: it's a bit mess, let's introduce separate tests

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImpl.java [hunk 2]
```diff
@@ -71,6 +77,10 @@ public CompletionStage<BatchedResult<T>> requestNextAsync(int rows) {
                     t = t.getCause();
                 }
 
+                if (tx != null) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImpl.java:80]: why we rollback the TX?

Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImpl.java:80]: we need to somehow handle tx here, this is exceptionally situation, i agree that probably not in all cases we need rollback, but this need to be discussed somehow and probably under differ issue.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java (hunk 1 above)
Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java:211]: done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImpl.java (hunk 2 above)
Comment (AMashenkov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImpl.java:80]: What are the cases when transaction must or shouldn't be rolled back?

One the one side. Here, as we already have a cursor, the query is already parsed and mapped to nodes.
So, we can't be sure whether any lock was obtained or not.

On the other side, it is a client API part and server errors are processed on the server-side.
Errors on the client side (e.g. disconnect and marshaller failures) can be handle by the user,
However, I doubt any error is recoverable here and TX will be likely aborted by the user.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientSqlTest.java [hunk 3]
```diff
@@ -121,6 +121,11 @@ void testExecuteAsyncDdlDml() {
             assertThrows(NoRowSetExpectedException.class, insertRes::currentPage);
         }
 
+        Transaction tx = client().transactions().begin();
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientSqlTest.java]: done

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlSynchronousApiTest.java [hunk 4]
```diff
@@ -159,10 +167,20 @@ public void dml() {
         IgniteSql sql = igniteSql();
```
Comment (korlov42) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlSynchronousApiTest.java:167]: Could you please add the following case for both Sync and Async embedded API implementation:
- start explicit transaction
- insert some rows to a table
- rollback transaction
- verify rows were not inserted

Comment (korlov42) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlSynchronousApiTest.java:167]: Oh, I see there is something similar in implicitTransactionsWithDml test. Let's split this onto several smaller tests

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 5]
```diff
@@ -366,7 +370,7 @@ public CompletableFuture<Void> closeAsync() {
                     );
                 });
 
-        stage.whenComplete((cur, ex) -> {
+        stage = stage.whenComplete((cur, ex) -> {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:373]: the idea behind is that you return future with attached cancellation handler to the client. Now you return the future of the handler to the client, and this doesn't make any sense

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionContext.java [hunk 6]
```diff
@@ -332,6 +339,18 @@ public interface RunnableX {
         void run() throws Throwable;
     }
 
+    /** Transaction for current context. */
+    public InternalTransaction transaction() {
+        return tx;
+    }
+
+    /** Transaction. */
+    public void transaction(InternalTransaction tx) {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionContext.java:348]: do we really need such setter? Looks like it is not used anywhere

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 7]
```diff
@@ -424,7 +436,7 @@ private void onError(RemoteException ex) {
 
         private void onNodeLeft(String nodeId) {
             remoteFragmentInitCompletion.entrySet().stream().filter(e -> nodeId.equals(e.getKey().nodeId()))
-                    .forEach(e -> e.getValue().completeExceptionally(new IgniteInternalException("asddd")));
+                    .forEach(e -> e.getValue().completeExceptionally(new IgniteInternalException("Node left, nodeId=" + nodeId)));
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:439]:
```suggestion
                    .forEach(e -> e.getValue().completeExceptionally(new IgniteInternalException("Node left the cluster [nodeId=" + nodeId + ']')));
```

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/MockedStructuresTest.java [hunk 8]
```diff
@@ -264,6 +273,48 @@ void before() throws Exception {
                 .get(1, TimeUnit.SECONDS);
     }
 
+    /**
+     * Checks inner transactions are initialized correctly.
+     */
+    @Test
+    public void testInnerTxInitiated() throws Exception {
+        SessionId sesId = queryProc.createSession(new PropertiesHolder() {
+            @Override
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/MockedStructuresTest.java:282]: `PropertiesHolder.holderFor(Map.of())`

**Comment by zstan** [comment]
@korlov42 is it ok for now ?

**Review by korlov42 — APPROVED** [review]

**Pull request closed without merging**